juniper/src/executor/look_ahead_lazy.rs

```diff
@@ -0,0 +1,1561 @@
+use std::collections::HashMap;
+
+use crate::{
+    ast::{Arguments, Directive, Field, Fragment, InputValue, Selection},
+    parser::{Span, Spanning},
+    value::ScalarValue,
+};
+
+use super::Variables;
+
+/// An enum that describes if a field is available in all types of the interface
+/// or only in a certain subtype
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Applies<'a> {
+    /// The field is available independent from the type
+    All,
+    /// The field is only available for a given typename
+    OnlyType(&'a str),
+}
+
+/// Shortcut for a [`Spanning`] containing a borrowed [`Span`].
+type BorrowedSpanning<'a, T> = Spanning<T, &'a Span>;
+
+/// JSON-like value that can be used as an argument in the query execution.
+///
+/// In contrast to an [`InputValue`], these values do only contain constants,
+/// meaning that variables get automatically resolved.
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[allow(missing_docs)]
+pub enum LookAheadValue<'a, S: ScalarValue + 'a> {
+    Null,
+    Scalar(&'a S),
+    Enum(&'a str),
+    List(LookAheadList<'a, S>),
+    Object(LookAheadObject<'a, S>),
+}
+
+impl<'a, S: ScalarValue + 'a> LookAheadValue<'a, S> {
+    fn from_input_value(
+        input_value: &'a InputValue<S>,
+        span: &'a Span,
```

Comment:
Mainly because `input_value` and `span` do not always come from the same source (because of `InputValue::Variable`, where this is re-entrant). Changing the argument to `BorrowedSpanning` will look less elegant at some callsites. Just keep them as two separate parameters. Also this is not an public function so i guess this isn't a big issue.

Reply:
@audunhalland well, the only place where they come not from the same source is [mapping an ` `InputValue::Variable`](https://github.com/graphql-rust/juniper/blob/149e61f3b52226121df03f43f3cd377315368b17/juniper/src/executor/look_ahead_lazy.rs#L49-L52), which is somewhat a special case regarding that we expand the variable. I think `BorrowedSpanning` cover the majority of cases better, and makes it much clearer that the provided `Span` is related to the `input_value`.

Reply:
Do you think the parameter should be a `BorrowedSpanning` or a `&BorrowedSpanning`? Sometimes we have only a `&BorrowedSpanning<InputValue>` that we want to map to a LookAheadValue. This is another reason it makes more sense to me to accept two individual borrowed parameters rather than one struct. The struct approach seems more shoe-horned.

Reply:
@audunhalland seems that just `BorrowedSpanning` works totally fine due to `Spanning::as_ref`: https://github.com/graphql-rust/juniper/pull/1212/commits/cecee4b3cc81596ec16e61c63ac99589c69c3c4d